File: app/src/main/java/tech/pauly/findapet/shared/events/ActivityEvent.java
package tech.pauly.findapet.shared.events;

import java.util.Objects;

public class ActivityEvent extends BaseViewEvent {

    private Class startActivityClass;
    private boolean finishActivity = false;

    public static ActivityEvent build(Object emitter) {
        return new ActivityEvent(emitter.getClass());
    }

    private ActivityEvent(Class emitter) {
        this.emitter = emitter;
    }

    public ActivityEvent finishActivity() {
        this.finishActivity = true;
        return this;
    }

    public ActivityEvent startActivity(Class startActivityClass) {
        this.startActivityClass = startActivityClass;
        return this;
    }

    public Class getStartActivityClass() {
        return startActivityClass;
    }

    public boolean isFinishActivity() {
        return finishActivity;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        ActivityEvent that = (ActivityEvent) o;
        return finishActivity == that.finishActivity && Objects.equals(startActivityClass, that.startActivityClass);
    }

    @Override
    public int hashCode() {

        return Objects.hash(startActivityClass, finishActivity);
    }
}

File: app/src/main/java/tech/pauly/findapet/shared/events/ActivityEvent.kt
package tech.pauly.findapet.shared.events

import android.content.Intent
import kotlin.reflect.KClass

open class ActivityEvent(private val emitter: Any,
                         open val startActivity: KClass<*>? = null,
                         open val finishActivity: Boolean = false,
                         open val customIntent: Intent? = null) : BaseViewEvent(emitter::class) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ActivityEvent

        if (emitter != other.emitter) return false
        if (startActivity != other.startActivity) return false
        if (finishActivity != other.finishActivity) return false
        if (customIntent != other.customIntent) return false

        return true
    }

    override fun hashCode(): Int {
        var result = emitter.hashCode()
        result = 31 * result + (startActivity?.hashCode() ?: 0)
        result = 31 * result + finishActivity.hashCode()
        result = 31 * result + (customIntent?.hashCode() ?: 0)
        return result
    }
}
